src/core/types.h

```diff
@@ -788,7 +788,8 @@ int binarySearch(T needle, RandomAccessIterator start, RandomAccessIterator end,
     int l = 0;
     int r = end - start - 1;
     while (l <= r) {
-        int mid = l + (r - l) / 2;
```

Comment:
Yeah, thanks for your recommended reference, and your solution is absolutely right!  But just as Joshua Bloch said in your recommended reference, @Daetalus 's thought about "int mid = low + ((high - low) / 2);" is right, at least for a right BinarySearch implementation. After analyzing your recommended article throughly and thinking about your mentioned overflow carefully, your expected reference may be one of the references of your reference -- Salvatore Ruggieri's paper about the average of two numbers of arbitrary sign. After I scanned that paper roughly, our BinarySearch problem is one special case of that article -- non-negative-division calculation (Section 3) where both l and r are non-negative and l <= r, so I think it also supports Daetalus's implementation.

Thanks again! 
Jianbin Wang